File: android/app/src/main/java/com/djs66256/short_drama/data/dto/RankingDramaDto.kt
package com.djs66256.short_drama.data.dto

import com.djs66256.short_drama.domain.model.RankingContentType
import com.djs66256.short_drama.domain.model.RankingDrama
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class RankingDramaDto(
    val id: String,
    val title: String,
    val description: String,
    @SerialName("cover_url")
    val coverUrl: String? = null,
    val category: String,
    @SerialName("episode_count")
    val episodeCount: Int,
    val tags: List<String> = emptyList(),
    val rating: Double = 0.0,
    @SerialName("created_at")
    val createdAt: String,
    @SerialName("updated_at")
    val updatedAt: String,
    @SerialName("content_type")
    val contentType: String,
    @SerialName("play_count")
    val playCount: Int,
    @SerialName("booking_count")
    val bookingCount: Int,
    @SerialName("recommendation_score")
    val recommendationScore: Double,
    @SerialName("is_booked")
    val isBooked: Boolean = false,
) {
    fun toDomain(): RankingDrama = RankingDrama(
        id = id,
        title = title,
        description = description,
        coverUrl = coverUrl.orEmpty(),
        category = category,
        episodeCount = episodeCount,
        tags = tags,
        rating = rating,
        createdAt = createdAt,
        updatedAt = updatedAt,
        contentType = RankingContentType.fromApiValue(contentType),
        playCount = playCount,
        bookingCount = bookingCount,
        recommendationScore = recommendationScore,
        isBooked = isBooked,
    )
}

File: ios/ShortDrama/Sources/Data/DTOs/RankingDramaDTO.swift
import Foundation

/// DTO for ranking list items.
struct RankingDramaDTO: Codable, Equatable {
    let id: String
    let title: String
    let description: String?
    let coverUrl: String?
    let category: String?
    let episodeCount: Int?
    let tags: [String]?
    let rating: Double?
    let createdAt: String?
    let updatedAt: String?
    let contentType: RankingContentTypeDTO?
    let playCount: Int?
    let bookingCount: Int?
    let recommendationScore: Double?
    let isBooked: Bool?
}

enum RankingContentTypeDTO: String, Codable, Equatable {
    case liveAction = "live_action"
    case ai

    func toEntity() -> RankingContentType {
        switch self {
        case .liveAction:
            return .liveAction
        case .ai:
            return .ai
        }
    }

    static func fallback(from category: String) -> RankingContentTypeDTO {
        let normalized = category.lowercased()
        if normalized.contains("ai") || normalized.contains("动画") || normalized.contains("动漫") || normalized.contains("漫剧") {
            return .ai
        }
        return .liveAction
    }
}

extension RankingDramaDTO {
    func toEntity() -> RankingDrama {
        let normalizedCategory = category?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let resolvedContentType = contentType?.toEntity()
            ?? RankingContentTypeDTO.fallback(from: normalizedCategory).toEntity()

        return RankingDrama(
            id: id,
            title: title,
            description: description?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "",
            coverUrl: coverUrl ?? "",
            category: normalizedCategory,
            episodeCount: episodeCount ?? 0,
            tags: tags?.filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty },
            rating: rating,
            createdAt: createdAt ?? "",
            updatedAt: updatedAt ?? "",
            contentType: resolvedContentType,
            playCount: playCount ?? 0,
            bookingCount: bookingCount ?? 0,
            recommendationScore: recommendationScore ?? 0,
            isBooked: isBooked ?? false
        )
    }
}
